tools/cli/internal/openapi/filter/hidden_envs.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2024 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package filter
+
+import (
+	"log"
+	"strings"
+
+	"github.com/getkin/kin-openapi/openapi3"
+)
+
+const (
+	hiddenEnvsExtensionName     = "x-xgen-hidden-env"
+	hiddenEnvsExtionsionKeyName = "envs"
+)
+
+type HiddenEnvsFilter struct{}
+
+func (f *HiddenEnvsFilter) Apply(oas *openapi3.T, metadata *Metadata) error {
+	for _, pathItem := range oas.Paths.Map() {
+		if err := applyOnPath(pathItem, metadata); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func applyOnPath(pathItem *openapi3.PathItem, metadata *Metadata) error {
+	for k, operation := range pathItem.Operations() {
+		if isOperationHiddenForEnv := isOperationHiddenForEnv(operation, metadata); isOperationHiddenForEnv {
+			log.Printf("Removing operation: '%s' from path: '%s' because is hidden for target env: %s", k, pathItem.Ref, metadata.targetEnv)
+			pathItem.SetOperation(k, nil) // Remove Operation if it is hidden for the target environment
+			continue
+		} else if operation.Extensions != nil {
+			// Remove the Hidden extension from the final OAS
+			delete(operation.Extensions, hiddenEnvsExtensionName)
+		}
+
+		for k, response := range operation.Responses.Map() {
+			if isResponseHiddenForEnv := isResponseHiddenForEnv(response, metadata); isResponseHiddenForEnv {
+				log.Printf("Removing response: '%s' from operationID: '%s' because is hidden for target env: %s", k, operation.OperationID, metadata.targetEnv)
+				operation.Responses.Delete(k) // Remove Response if it is hidden for the target environment
+			} else if response.Extensions != nil {
+				// Remove the Hidden extension from the final OAS
+				delete(response.Extensions, hiddenEnvsExtensionName)
+			}
+		}
+
+		if isRequestBodyHiddenForEnv := isRequestBodyHiddenForEnv(operation.RequestBody, metadata); isRequestBodyHiddenForEnv {
+			log.Printf("Removing requestBody from operationID: '%s' because is hidden for target env: %s", operation.OperationID, metadata.targetEnv)
+			operation.RequestBody = nil // Remove RequestBody if it is hidden for the target environment
+		} else if operation.RequestBody != nil && operation.RequestBody.Extensions != nil {
+			// Remove the Hidden extension from the final OAS
+			delete(operation.RequestBody.Extensions, hiddenEnvsExtensionName)
+		}
```

Comment:
can we try to have two separate funcs hideResponses and hideRequestBody or updateResponses/updateRequestBody

Reply:
updated